pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,226 @@
+			zoneToInstanceTypeInfo[zone] = instanceTypes.([]*ec2.InstanceTypeInfo)
+			continue
+		}
+
+		// populate the cache by zonal keys
+		instanceTypes, err := p.getAllInstanceTypes(ctx)
+		if err != nil {
+			return nil, err
+		}
+		instanceTypes, err = p.filterByLocation(ctx, instanceTypes, zone)
+		if err != nil {
+			return nil, err
+		}
+		p.instanceTypeCache.SetDefault(zone, instanceTypes)
+		zoneToInstanceTypeInfo[zone] = instanceTypes
+	}
+
+	for zone, instanceTypes := range zoneToInstanceTypeInfo {
+		zoneToInstanceTypeInfo[zone] = p.filterFrom(instanceTypes, constraints)
+	}
+	return zoneToInstanceTypeInfo, nil
+}
+
+// UniqueInstanceTypesFrom returns a unique slice of InstanceTypeInfo structs from the values of the instancePools passed in
+func (p *InstanceTypeProvider) UniqueInstanceTypesFrom(instancePools map[string][]*ec2.InstanceTypeInfo) []*ec2.InstanceTypeInfo {
+	result := []*ec2.InstanceTypeInfo{}
+	uniqueInstanceTypes := map[string]bool{}
+	for _, instanceTypes := range instancePools {
+		for _, instanceType := range instanceTypes {
+			if _, ok := uniqueInstanceTypes[*instanceType.InstanceType]; !ok {
+				uniqueInstanceTypes[*instanceType.InstanceType] = true
+				result = append(result, instanceType)
+			}
+		}
+	}
+	return result
+}
+
+// InstanceTypesPerZoneFrom returns a mapping of zone to InstanceTypeInfo based on the provided slice of instance types names and mapping of all zones to InstanceTypeInfo
+func (p *InstanceTypeProvider) InstanceTypesPerZoneFrom(instanceTypeNames []string, instancePools map[string][]*ec2.InstanceTypeInfo) map[string][]*ec2.InstanceTypeInfo {
+	result := map[string][]*ec2.InstanceTypeInfo{}
+	for zone, instanceTypes := range instancePools {
+		for _, instanceTypeName := range instanceTypeNames {
+			for _, instanceType := range instanceTypes {
+				if instanceType != nil && *instanceType.InstanceType == instanceTypeName {
+					result[zone] = append(result[zone], instanceType)
+				}
+			}
+		}
+	}
+	return result
+}
+
+// getAllInstanceTypes retrieves all instance types from the ec2 DescribeInstanceTypes API using some opinionated filters
+func (p *InstanceTypeProvider) getAllInstanceTypes(ctx context.Context) ([]*ec2.InstanceTypeInfo, error) {
+	instanceTypes := []*ec2.InstanceTypeInfo{}
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("supported-virtualization-type"),
+				Values: []*string{aws.String("hvm")},
+			},
+			{
+				Name:   aws.String("bare-metal"),
+				Values: []*string{aws.String("false")},
+			},
+			{
+				Name:   aws.String("instance-type"),
+				Values: aws.StringSlice([]string{"m5.*", "c5.*", "r5.*", "m5a.*", "c5a.*", "r5a.*", "m4.*", "c4.*", "r4.*", "m6g.*", "c6g.*", "r6g.*"}),
+			},
+		},
+	}
+	err := p.ec2.DescribeInstanceTypesPagesWithContext(ctx, describeInstanceTypesInput, func(page *ec2.DescribeInstanceTypesOutput, lastPage bool) bool {
```

Comment:
In the other providers, we drop a debug log on every single EC2 API call so that it's apparant the number of calls were making. Do you think it's worthwhile to continue this pattern here?